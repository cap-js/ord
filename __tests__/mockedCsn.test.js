const cds = require("@sap/cds");
const path = require("path");
const { AUTHENTICATION_TYPE } = require("../lib/constants");

describe("Tests for ORD document generated out of mocked csn files", () => {
    let ord;

    function checkOrdDocumentInternal(csn) {
        const document = ord(csn);

        expect(document).not.toBeUndefined();
        expect(document.packages).toBeDefined();
        expect(document.apiResources).toHaveLength(3);
        expect(document.eventResources).toHaveLength(3);
    }

    function checkOrdDocumentPrivate(csn) {
        const document = ord(csn);

        expect(document).not.toBeUndefined();
        expect(document.packages).toHaveLength(0);
        expect(document.apiResources).toHaveLength(0);
        expect(document.eventResources).toHaveLength(0);
    }

    beforeAll(() => {
        cds.root = path.join(__dirname, "bookshop");
        jest.spyOn(cds, "context", "get").mockReturnValue({
            authConfig: {
                types: [AUTHENTICATION_TYPE.Open],
            },
        });
        jest.spyOn(require("../lib/date"), "getRFC3339Date").mockReturnValue("2024-11-04T14:33:25+01:00");
        ord = require("../lib/ord");
    });

    beforeEach(() => {
        cds.env.ord = {
            namespace: "sap.test.cdsrc.sample",
            openResourceDiscovery: "1.10",
            description: "this is my custom description",
            policyLevels: ["sap:core:v1"],
        };
    });

    afterAll(() => {
        jest.restoreAllMocks();
        jest.clearAllMocks();
    });

    describe("Tests for ORD document when .cdsrc.json has no `ord` property", () => {
        test("Successfully create ORD Documents with no `ord` in .cdsrc.json", () => {
            cds.env = {};
            const csn = require("./__mocks__/publicResourcesCsn.json");
            const document = ord(csn);
            expect(document).not.toBeUndefined();
            expect(document).toMatchSnapshot();
        });
    });

    describe("Tests for ORD document when there no events or entities in service definitions", () => {
        test("Successfully create ORD Document: no entityTypeMappings field in apiResource", () => {
            cds.env = {};
            const csn = require("./__mocks__/noEntitiesInServiceDefinitionCsn.json");
            const document = ord(csn);

            expect(document).not.toBeUndefined();
            expect(document.apiResources).toHaveLength(1);
            expect(document.eventResources).toHaveLength(1);
            expect(document.apiResources[0].ordId).toEqual(expect.stringContaining("EbMtEmitter"));
            expect(document.eventResources[0].ordId).toEqual(expect.stringContaining("EbMtEmitter"));
            expect(document.apiResources[0].entityTypeMappings).toBeUndefined();
        });
    });

    describe("Tests for ORD document checking if entityTypes and entityTypeMappings are generated correctly", () => {
        test("Successfully create ORD Document: entityTypes with local entities and entityTypeMappings containing referenced entities", () => {
            const csn = require("./__mocks__/localAndNonODMReferencedEntitiesCsn.json");
            const document = ord(csn);

            expect(document).not.toBeUndefined();
            expect(document.entityTypes).toHaveLength(2);
            expect(document.entityTypes[0].partOfPackage).toEqual(expect.stringContaining("entityType"));
            expect(document.entityTypes[0].level).toEqual(expect.stringContaining("root-entity"));
            expect(document.apiResources[0].entityTypeMappings[0].entityTypeTargets).toEqual(
                expect.arrayContaining([
                    { ordId: "sap.odm:entityType:SomeODMEntity:v1" },
                    { ordId: "sap.sm:entityType:SomeAribaDummyEntity:v1" },
                ]),
            );
        });
    });

    describe("Tests for ORD document when all the resources are private", () => {
        test("All services are private: Successfully create ORD Documents without packages, empty apiResources and eventResources lists", () => {
            const csn = require("./__mocks__/privateResourcesCsn.json");
            checkOrdDocumentPrivate(csn);
        });
    });

    describe("Tests for ORD document when all the resources are internal", () => {
        test("All services are internal: Successfully create ORD Documents without packages, empty apiResources and eventResources lists", () => {
            const csn = require("./__mocks__/internalResourcesCsn.json");
            checkOrdDocumentInternal(csn);
        });
    });

    describe("Tests for ORD document when there no events or entities in service definitions", () => {
        test("Successfully create ORD Documents: no Catalog service in apiResource; no Admin service in eventResources", () => {
            const csn = require("./__mocks__/noApisOrNoEventsCsn.json");
            const document = ord(csn);

            expect(document).not.toBeUndefined();
            expect(document.apiResources).toHaveLength(2);
            expect(document.eventResources).toHaveLength(2);
            expect(document.apiResources[0].ordId).toEqual(expect.stringContaining("AdminService"));
            expect(document.eventResources[1].ordId).toEqual(expect.stringContaining("CatalogService"));
        });
    });

    describe("Tests for ORD document when service is annotated as a primary Data Product`", () => {
        test("Successfully create ORD Documents: ", () => {
            const csn = require("./__mocks__/dataProductCsn.json");
            const document = ord(csn);

            expect(document).not.toBeUndefined();
            expect(document.apiResources).toHaveLength(2);
            const dataProductApiResources = document.apiResources.filter(
                (resource) => resource.implementationStandard === "sap.dp:data-subscription-api:v1",
            );
            expect(dataProductApiResources).toHaveLength(2);
            expect(dataProductApiResources[0].resourceDefinitions).toHaveLength(1);
            expect(dataProductApiResources[0].resourceDefinitions[0]).type === "sap-csn-interop-effective-v1";
        });
    });
});
