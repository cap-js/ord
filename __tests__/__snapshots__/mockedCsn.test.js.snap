// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for ORD document generated out of mocked csn files Tests for ORD document when .cdsrc.json has no \`ord\` property Successfully create ORD Documents with no \`ord\` in .cdsrc.json 1`] = `
{
  "$schema": "https://open-resource-discovery.github.io/specification/spec-v1/interfaces/Document.schema.json",
  "apiResources": [
    {
      "apiProtocol": "odata-v4",
      "description": "Description for AdminService",
      "entityTypeMappings": {
        "entityTypeTargets": [
          {
            "ordId": "sap.odm:entityType:test-from-extension:v1",
          },
        ],
      },
      "entryPoints": [
        "/odata/v4/admin",
      ],
      "exposedEntityTypes": [
        {
          "ordId": "sap.odm:entityType:odm.bookshop.Authors:v1",
        },
      ],
      "extensible": {
        "supported": "yes",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:apiResource:AdminService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:AdminService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:AdminService:v1/AdminService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:AdminService:v1/AdminService.edmx",
        },
      ],
      "shortDescription": "short description for test AdminService",
      "title": "This is test AdminService title",
      "version": "2.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for CatalogService",
      "entryPoints": [
        "/browse",
      ],
      "extensible": {
        "supported": "yes",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:apiResource:CatalogService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:CatalogService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:CatalogService:v1/CatalogService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:CatalogService:v1/CatalogService.edmx",
        },
      ],
      "shortDescription": "short description for test CatalogService",
      "title": "This is test Catalog Service title",
      "version": "2.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for CinemaService",
      "entryPoints": [
        "/browse",
      ],
      "extensible": {
        "supported": "yes",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:apiResource:CinemaService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:CinemaService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:CinemaService:v1/CinemaService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:CinemaService:v1/CinemaService.edmx",
        },
      ],
      "shortDescription": "short description for test CinemaService",
      "title": "This is test Cinema Service title",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "consumptionBundles": [
    {
      "description": "This Consumption Bundle contains all resources of the reference app which are unprotected and do not require authentication",
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "capirebookshopordsample:consumptionBundle:noAuth:v1",
      "shortDescription": "If we have another protected API then it will be another object",
      "title": "Unprotected resources",
      "version": "1.0.0",
    },
  ],
  "description": "this is an application description",
  "eventResources": [
    {
      "description": "CAP Event resource describing events / messages.",
      "entityTypeMappings": {
        "entityTypeTargets": [
          {
            "ordId": "sap.odm:entityType:test-from-extension:v1",
          },
        ],
      },
      "exposedEntityTypes": [
        {
          "ordId": "sap.odm:entityType:odm.bookshop.Authors:v1",
        },
      ],
      "extensible": {
        "supported": "yes",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:eventResource:AdminService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:AdminService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/ord/v1/customer.capirebookshopordsample:eventResource:AdminService:v1/AdminService.asyncapi2.json",
        },
      ],
      "shortDescription": "short description for test AdminService",
      "title": "This is test AdminService title",
      "version": "2.0.0",
      "visibility": "public",
    },
    {
      "description": "CAP Event resource describing events / messages.",
      "extensible": {
        "supported": "yes",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:eventResource:CatalogService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:CatalogService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/ord/v1/customer.capirebookshopordsample:eventResource:CatalogService:v1/CatalogService.asyncapi2.json",
        },
      ],
      "shortDescription": "short description for test CatalogService",
      "title": "This is test Catalog Service title",
      "version": "2.0.0",
      "visibility": "public",
    },
    {
      "description": "CAP Event resource describing events / messages.",
      "extensible": {
        "supported": "yes",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:eventResource:CinemaService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:CinemaService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/ord/v1/customer.capirebookshopordsample:eventResource:CinemaService:v1/CinemaService.asyncapi2.json",
        },
      ],
      "shortDescription": "short description for test CinemaService",
      "title": "This is test Cinema Service title",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "groups": [
    {
      "groupId": "sap.cds:service:customer.capirebookshopordsample:AdminService",
      "groupTypeId": "sap.cds:service",
      "title": "This is test AdminService title",
    },
    {
      "groupId": "sap.cds:service:customer.capirebookshopordsample:CatalogService",
      "groupTypeId": "sap.cds:service",
      "title": "This is test Catalog Service title",
    },
    {
      "groupId": "sap.cds:service:customer.capirebookshopordsample:CinemaService",
      "groupTypeId": "sap.cds:service",
      "title": "This is test Cinema Service title",
    },
  ],
  "openResourceDiscovery": "1.9",
  "packages": [
    {
      "description": "This package contains public General for capire bookshop ord sample.",
      "ordId": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "shortDescription": "Package containing public General",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "1.0.0",
    },
  ],
  "policyLevels": [
    "none",
  ],
  "products": [
    {
      "ordId": "customer:product:capire.bookshop.ord.sample:",
      "shortDescription": "Short description of capire bookshop ord sample",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
    },
  ],
}
`;

exports[`Tests for ORD document generated out of mocked csn files Tests for ORD document when there no events or entities in service definitions Successfully create ORD Document: no entityTypeMappings field in apiResource 1`] = `
{
  "$schema": "https://open-resource-discovery.github.io/specification/spec-v1/interfaces/Document.schema.json",
  "apiResources": [
    {
      "apiProtocol": "odata-v4",
      "description": "Description for EbMtEmitter",
      "entryPoints": [
        "/odata/v4/eb-mt-emitter",
      ],
      "extensible": {
        "supported": "no",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:apiResource:EbMtEmitter:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:EbMtEmitter",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:EbMtEmitter:v1/EbMtEmitter.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/customer.capirebookshopordsample:apiResource:EbMtEmitter:v1/EbMtEmitter.edmx",
        },
      ],
      "shortDescription": "Short description of EbMtEmitter",
      "title": "EbMtEmitter",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "consumptionBundles": [
    {
      "description": "This Consumption Bundle contains all resources of the reference app which are unprotected and do not require authentication",
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "capirebookshopordsample:consumptionBundle:noAuth:v1",
      "shortDescription": "If we have another protected API then it will be another object",
      "title": "Unprotected resources",
      "version": "1.0.0",
    },
  ],
  "description": "this is an application description",
  "eventResources": [
    {
      "description": "CAP Event resource describing events / messages.",
      "extensible": {
        "supported": "no",
      },
      "lastUpdate": "2024-11-04T14:33:25+01:00",
      "ordId": "customer.capirebookshopordsample:eventResource:EbMtEmitter:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:EbMtEmitter",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/ord/v1/customer.capirebookshopordsample:eventResource:EbMtEmitter:v1/EbMtEmitter.asyncapi2.json",
        },
      ],
      "shortDescription": "EbMtEmitter event resource",
      "title": "ODM capirebookshopordsample Events",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "groups": [
    {
      "groupId": "sap.cds:service:customer.capirebookshopordsample:EbMtEmitter",
      "groupTypeId": "sap.cds:service",
      "title": "EbMtEmitter Service",
    },
  ],
  "openResourceDiscovery": "1.9",
  "packages": [
    {
      "description": "This package contains public General for capire bookshop ord sample.",
      "ordId": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "shortDescription": "Package containing public General",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "1.0.0",
    },
  ],
  "policyLevels": [
    "none",
  ],
  "products": [
    {
      "ordId": "customer:product:capire.bookshop.ord.sample:",
      "shortDescription": "Short description of capire bookshop ord sample",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
    },
  ],
}
`;
